arch/arm/core/thread.c

```diff
@@ -56,22 +56,53 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 		 int priority, unsigned int options)
 {
 	char *pStackMem = K_THREAD_STACK_BUFFER(stack);
+	char *stackEnd;
+	/* Offset between the top of stack and the high end of stack area. */
+	u32_t top_of_stack_offset = 0;
 
 	_ASSERT_VALID_PRIO(priority, pEntry);
 
+#if defined(CONFIG_USERSPACE)
+	/* Truncate the stack size to align with the MPU region granularity.
+	 * This is done proactively to account for the case when the thread
+	 * switches to user mode (thus, its stack area will need to be MPU-
+	 * programmed to be assigned unprivileged RW access permission).
+	 */
+	stackSize &= ~(CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE - 1);
+
+#ifdef CONFIG_THREAD_USERSPACE_LOCAL_DATA
+	/* Reserve space on top of stack for local data. */
+	u32_t p_local_data = STACK_ROUND_DOWN(pStackMem + stackSize
+		- sizeof(*thread->userspace_local_data));
+
+	thread->userspace_local_data =
+		(struct _thread_userspace_local_data *)(p_local_data);
+
+	/* Top of actual stack must be moved below the user local data. */
+	top_of_stack_offset = (u32_t)
+		(pStackMem + stackSize - ((char *)p_local_data));
+
+#endif /* CONFIG_THREAD_USERSPACE_LOCAL_DATA */
+#endif /* CONFIG_USERSPACE */
+
 #if CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT && CONFIG_USERSPACE
-	char *stackEnd = pStackMem + stackSize - MPU_GUARD_ALIGN_AND_SIZE;
+	/* This is required to work-around the case where the thread
```

Comment:
@andrewboie this work-around is "kept" here, but it is only needed since we do not call k_thread_create() with size argument coming from K_THREAD_STACK_SIZEOF. I believe we need to fix this; enforce all calls to k_thread_create() have the stack parameter taken from K_THREAD_STACK_SIZEOF(x), where x = K_THREAD_STACK_DEFINED(<some size>)